stream.go

```diff
@@ -140,6 +144,198 @@ func (r *Client) StreamPrediction(ctx context.Context, prediction *Prediction) (
+			return
+		}
+		lastEventID = event.Id()
+		switch event.Event() {
+		case SSETypeOutput:
+			io.WriteString(writer, event.Data())
+		case SSETypeDone:
+			writer.Close()
+			return
+		case SSETypeLogs:
+			// TODO
+		default:
+			writer.CloseWithError(fmt.Errorf("unknown event type %s", event.Event()))
+			return
+		}
+	}
+}
+
+type dataURL struct {
+	url string
+}
+
+var _ streaming.File = &dataURL{}
+
+func (d *dataURL) Body(_ context.Context) (io.ReadCloser, error) {
+	data, err := dataurl.DecodeString(d.url)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return io.NopCloser(bytes.NewReader(data.Data)), nil
+}
+
+type httpURL struct {
+	c   *http.Client
+	url string
+}
+
+var _ streaming.File = &httpURL{}
+
+func (h *httpURL) Body(ctx context.Context) (io.ReadCloser, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, h.url, nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := h.c.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	return resp.Body, nil
+}
+
+type errWrapper struct {
+	err error
+}
+
+var _ streaming.File = &errWrapper{}
+
+func fileError(err error) streaming.File {
+	return &errWrapper{err: err}
+}
+
+func (e *errWrapper) Body(_ context.Context) (io.ReadCloser, error) {
+	return nil, e.err
+}
+
+func (e *errWrapper) Close() error {
+	return nil
+}
+
+func (r *Client) streamFilesTo(ctx context.Context, out chan<- streaming.File, url string, lastEventID string) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		out <- fileError(err)
+		close(out)
+		return
+	}
+	req.Header.Set("Accept", "text/event-stream")
+	req.Header.Set("Cache-Control", "no-cache")
+	req.Header.Set("Connection", "keep-alive")
+
+	if lastEventID != "" {
+		req.Header.Set("Last-Event-ID", lastEventID)
+	}
+
+	resp, err := r.c.Do(req)
+	if err != nil {
+		out <- fileError(fmt.Errorf("failed to send request: %w", err))
+		close(out)
+		return
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		out <- fileError(fmt.Errorf("received invalid status code: %d", resp.StatusCode))
+		close(out)
+		return
+	}
+	defer resp.Body.Close()
+	decoder := eventsource.NewDecoder(resp.Body)
+	for {
+		event, err := decoder.Decode()
+		if err != nil {
+			if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
+				// retry (TODO: backoff policy?)
+				r.streamFilesTo(ctx, out, url, lastEventID)
+				return
+			}
+			out <- fileError(fmt.Errorf("Failed to get token: %w", err))
+			close(out)
+			return
+		}
+		lastEventID = event.Id()
+		switch event.Event() {
+		case SSETypeOutput:
+			if strings.HasPrefix(event.Data(), "data:") {
+				out <- &dataURL{url: event.Data()}
+			} else if strings.HasPrefix(event.Data(), "http") {
+				out <- &httpURL{c: r.c, url: event.Data()}
+			}
+		case SSETypeDone:
+			close(out)
+			return
+		case SSETypeLogs:
+			// TODO
+		default:
+			out <- fileError(fmt.Errorf("unknown event type %s", event.Event()))
```

Comment:
I think we forgot to `close(out)` here? Which possibly makes my case about doing this once stronger?